### PR TITLE
Added "open folder" function to asset explorer. Generalized the Panel…

### PR DESCRIPTION
**Pull request by eanders-MS**
… component, and allow user to inject controls into its header. Added placeholder ConversationEditor.

### CODE REVIEW

File: client-v2/src/constants.js [hunk 1]
```diff
@@ -1,7 +1,8 @@
```
Comment (Fignewton5) [inline client-v2/src/constants.js]: Copyright needed

File: client-v2/src/ui/editor/botChatEditor/chatPanel.js [hunk 2]
```diff
@@ -45,8 +44,10 @@ export default class ChatPanel extends React.Component {
     render() {
         return (
             <div className={ CSS }>
-                <Panel>
-                    <Chat />
+                <Panel title='Chat'>
```
Comment (compulim) [inline client-v2/src/ui/editor/botChatEditor/chatPanel.js]: Use double quote.

```jsx
<Panel title="Chat">
```

File: client-v2/src/ui/editor/botChatEditor/detailPanel.js [hunk 3]
```diff
@@ -45,8 +44,10 @@ export default class DetailPanel extends React.Component {
     render() {
         return (
             <div className={ CSS }>
-                <Panel>
-                    <Detail />
+                <Panel title='Details'>
```
Comment (compulim) [inline client-v2/src/ui/editor/botChatEditor/detailPanel.js]: Use double quotes

```jsx
<Panel title="Details">
```

File: client-v2/src/ui/editor/botChatEditor/logPanel.js [hunk 4]
```diff
@@ -45,8 +44,10 @@ export default class LogPanel extends React.Component {
     render() {
         return (
             <div className={ CSS }>
-                <Panel>
-                    <Log />
+                <Panel title='Log'>
```
Comment (compulim) [inline client-v2/src/ui/editor/botChatEditor/logPanel.js]: Use double quotes.

```jsx
<Panel title="Log">
```

File: client-v2/src/ui/editor/conversationEditor/editorPanel.js [hunk 5]
```diff
@@ -34,36 +34,28 @@
...
+export default class EditorPanel extends React.Component {
     render() {
         return (
             <div className={ CSS }>
-                <div className={ HEADER_CSS }>
-                </div>
-                <div className={ BODY_CSS }>
-                    { this.props.children }
-                </div>
+                <Panel title='Editor'>
```
Comment (compulim) [inline client-v2/src/ui/editor/conversationEditor/editorPanel.js]: Use double quotes.

```jsx
<Panel title="Editor">
```

File: client-v2/src/ui/editor/conversationEditor/propertiesPanel.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+const CSS = css({
+    height: '100%'
+});
+
+export default class PropertiesPanel extends React.Component {
+    render() {
+        return (
+            <div className={ CSS }>
+                <Panel title='Properties'>
```
Comment (compulim) [inline client-v2/src/ui/editor/conversationEditor/propertiesPanel.js]: Use double quotes.

```jsx
<Panel title="Properties">
```

File: client-v2/src/ui/shell/explorer/assetExplorerSet.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+import CardExplorer from './cardExplorer';
+import LuisExplorer from './luisExplorer';
+import QnAExplorer from './qnaExplorer';
+import FormExplorer from './formExplorer';
+import ConversationExplorer from './conversationExplorer';
+import FolderNotOpenExplorer from './folderNotOpenExplorer';
+
+if (typeof window !== 'undefined') { require = window['require']; }
+
+const fs = require('fs');
```
Comment (compulim) [inline client-v2/src/ui/shell/explorer/assetExplorerSet.js]: Please add a `// TODO` here for me, we will try to move all native calls to the main thread. I am still working on the bridge. :)

```js
// TODO: Should move all native calls to main thread
```

File: client-v2/src/ui/shell/explorer/conversationExplorer.js [hunk 8]
```diff
@@ -55,19 +57,54 @@ const BOTS_CSS = css({
...
+                    </ExpandCollapse>
+                </li>
+            </ul>
+        );
+    }
+}
+
+export default connect(state => ({
+    folder: state.assetExplorer.folder,
+    documents: state.editor.documents.filter(doc => doc.contentType === constants.ContentType_Converation)
```
Comment (compulim) [inline client-v2/src/ui/shell/explorer/conversationExplorer.js]: This call to `Array.filter()` will create a new `documents` array every time `connect` is called. Two things to do, either:
1. Just pass documents inside the React component and process it there (in `render` or in `willReceiveProps` -> `setState`)
2. Use [`reselect`](https://npmjs.com/package/reselect) or any memorize function, `n=1` memoizer is okay

File: client-v2/src/ui/shell/explorer/folderNotOpenExplorer.js [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    constructor(props, context) {
+        super(props, context);
+
+        this.handleOpenFolderClick = this.handleOpenFolderClick.bind(this);
+    }
+
+    handleOpenFolderClick(e) {
+        e.stopPropagation();
+        dialog.showOpenDialog({
+                properties: ['openDirectory']
```
Comment (compulim) [inline client-v2/src/ui/shell/explorer/folderNotOpenExplorer.js]: Indentation.

File: client-v2/src/ui/shell/explorer/folderNotOpenExplorer.js [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+            },
+            (filepaths) => {
```
Comment (compulim) [inline client-v2/src/ui/shell/explorer/folderNotOpenExplorer.js]: No need parenthesis here.

```js
filepaths => {
```

File: client-v2/src/ui/shell/explorer/folderNotOpenExplorer.js [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+class FolderNotOpenExplorer extends React.Component {
+    constructor(props, context) {
+        super(props, context);
+
+        this.handleOpenFolderClick = this.handleOpenFolderClick.bind(this);
+    }
+
+    handleOpenFolderClick(e) {
+        e.stopPropagation();
+        dialog.showOpenDialog({
```
Comment (compulim) [inline client-v2/src/ui/shell/explorer/folderNotOpenExplorer.js]: Is `dialog` a global variable?

If we need to implement dialog, could consider [`react-augmentation`](https://npmjs.com/package/react-augmentation) for layer management.

File: client-v2/src/ui/shell/mdi/conversationTab.js [hunk 12]
```diff
@@ -0,0 +1,7 @@
+import React from 'react';
+import { connect } from 'react-redux';
+
```
Comment (compulim) [inline client-v2/src/ui/shell/mdi/conversationTab.js]: Remove extra line.

File: client-v2/src/ui/shell/explorer/folderNotOpenExplorer.js (hunk 11 above)
Comment (eanders-MS) [reply client-v2/src/ui/shell/explorer/folderNotOpenExplorer.js]: It's an object from require('electron').remote. Should be moved to the main thread. I've added a TODO in my PR feedback PR, here: https://github.com/Microsoft/BotFramework-Emulator/pull/390

**Pull request merged**